server/lib/nl/counters.py

```diff
@@ -0,0 +1,70 @@
+# Copyright 2023 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import time
+from typing import Dict
+
+
+#
+# A helper class to track info / error / timing counters.
+#
+class Counters:
+
+  def __init__(self):
+    self._warn: Dict = {}
+    self._info: Dict = {}
+    self._timing: Dict = {}
+
+  #
+  # Adds/updates an info counter.
+  #
+  # For a given counter, caller should always pass the same type
+  # for value.  If value is numeric, then its a single updated
+  # counter, otherwise, counter is a list of values.
+  #
+  def info(self, counter: str, value: any):
```

Comment:
reading the code below it seems that this is expected to either be an int or float or a List of values, right? In that case, a more useful type hint would be:

`def info(self, counter: str, value: Union[int, float, List]):` or even better if the List is expected to have either floats or ints too:

`def info(self, counter: str, value: Union[int, float, List[Union[int, float]]]):`

Reply:
I have some opinions about use of Union type in py :), and I admit it is a taste thing.

Imo the hard-to-read/long Union structure somehow doesn't seem to justify its presence, especially when python won't enforce it either.  I view  typing as a useful hint to know what type it can take (for IDE esp), and that's not too helpful when its Union which can be one of many.

And finally, the type can be more than beyond just int/float... it can be str, enums (ContainedInPlaceType), etc.